### DISCUSSION TITLE
Annotating OneOrMany errors?

### DISCUSSION THREAD

**Discussion by wez** [jonasbb/serde_with#779] [Q&A] [answered]
I just ran into an issue where I have:

```rust
#[serde_as(deserialize_as = "OneOrMany<_, PreferOne>")]
pub action: Vec<Action>,
```

and a user tried to set:

```json
{
  "action": [
   [ "Suspend" ]
 ] 
}
```

which is invalid because they are trying to assign an array of arrays of Action instead of just an array of Action.

The resulting error message looks like:

```
 OneOrMany could not deserialize any variant:
  One: invalid type: sequence, expected string or map
  Many: invalid type: sequence, expected string or map at line 263 column 1
```

line 263 is the end of the input json document.

In the interest of trying to improve the error message, I tried to apply `serde_path_to_error` to this, but it cannot see the path.

I think it would be nice if I could specialize or newtype-wrap that `OneOrMany` in some way and provide an override for the `expecting` method of the underlying visitor, so that I could have it say `either a single Action or an array of Action`.

I've tried to apply `DeserializeAsWrap` without success:

```rust
fn one_or_many<'de, T, D>(deserializer: D, expecting: &'static str) -> Result<Vec<T>, D::Error>
where
    T: Deserialize<'de> ,
    D: Deserializer<'de>,
{

    let result = serde_with::de::DeserializeAsWrap::<
        Vec<T>,
        OneOrMany<T, PreferOne>  // I'm unsure if I need T or Vec<T> here, but neither compile
    >::deserialize(deserializer)?; // the intent is to map_err here and do some editorializing with `expecting`

    Ok(result.into_inner())
}

fn one_or_many_action<'de, D>(deserializer: D) -> Result<Vec<Action>, D::Error>
where
    D: Deserializer<'de>,
{
    one_or_many(deserializer, "either a single Action or an array of Action, for field action")
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Hash, Clone)]
struct Rule {
  #[serde(deserialize_with = "one_or_many_action")]
  pub action: Vec<Action>,
}
```

but I think I'm having a similar naming/resolution issue as is mentioned in https://github.com/jonasbb/serde_with/discussions/649

I'm looking for a suggestion (or ideally, a solution!) for how to achieve an annotated/expanded error message for this sort of situation!

**Comment by jonasbb** [comment, marked as answer]
The error message you see is part of the `DeserializeAs` implementation of `OneOrMany`. As such it is not really possible to change it, as this would require modifying the `DeserializeAs` code. In the future maybe it might be possible to refer to a const generic static string, but that is impossible right now too.

https://github.com/jonasbb/serde_with/blob/aede9d569029927e9739492d103a7285e2fb90ea/serde_with/src/de/impls.rs#L1527-L1557

I think this matches what you are trying to do in the code snippet. It does not replace the error message of `OneOrMany`. Instead if any error that pops up is replaced with the custom error message. This might hide some errors that are coming from the original deserializer.

```rust
fn one_or_many<'de, T, D>(deserializer: D) -> Result<Vec<T>, D::Error>
where
    T: Deserialize<'de>,
    D: Deserializer<'de>,
{
    let result: Result<Vec<T>, _> = OneOrMany::<Same, formats::PreferOne>::deserialize_as(deserializer);
    match result {
        Ok(r) => Ok(r),
        Err(err) => Err(serde::de::Error::custom("My Custom Error Message"))
    }
}
```

**Reply by wez** [reply]
Thank you! 

This is what I ended up with:

```rust
fn one_or_many<'de, T, D>(deserializer: D, expecting: &str) -> Result<Vec<T>, D::Error>
where
    T: Deserialize<'de>,
    D: Deserializer<'de>,
{
    let result: Result<Vec<T>, _> =
        OneOrMany::<serde_with::Same, PreferOne>::deserialize_as(deserializer);
    match result {
        Ok(r) => Ok(r),
        Err(err) => Err(serde::de::Error::custom(format!(
            "{expecting}.\nThe underlying error message is:\n{err:#}"
        ))),
    }
}

fn one_or_many_action<'de, D>(deserializer: D) -> Result<Vec<Action>, D::Error>
where
    D: Deserializer<'de>,
{
    one_or_many(
        deserializer,
        "\"action\" field expected either a single Action or an array of Actions",
    )
}
```

Then in my struct:

```rust
#[serde(deserialize_with = "one_or_many_action")]
pub action: Vec<Action>,
```

Produces an error like this, which still isn't perfect, but at least points to the action field:

```
"action" field expected either a single Action or an array of Actions.
The underlying error message is:
OneOrMany could not deserialize any variant:
  One: invalid type: sequence, expected string or map
  Many: invalid type: sequence, expected string or map at line 263 column 1
```
